equality assertions

avikchaudhuri opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #102256

Previously we had runtime asserts for range constraints. This diff adds runtime asserts for equality constraints.

This requires a bit of refactoring that is worth calling out.
1. [Minor] Some of the data structures produced by export and consumed by the runtime assertion pass need to be broadened. This is a WIP. There are some associated code improvements that are included in this diff, but by and large the structures are similar to what exists now. Meanwhile @angelayi and I are chatting about how to make it qualitatively better: briefly, we want to index everything by symbols, which are 1-1 with (name, dim) pairs.
2. [Major] The order in which runtime asserts are emitted is changed. Previously we used to do the work in `placeholder`, now this diff adds a hook for "post-processing" after processing of all placeholders is done. This is needed because equality constraints can mention different placeholders. This change also opens the way to optimizing codegen: e.g., each (name, dim) pair should correspond to a single intermediate variable that is reused across runtime asserts. This is future work. 

Differential Revision: [D46177642](https://our.internmc.facebook.com/intern/diff/D46177642/)

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx

avikchaudhuri committed:
equality assertions



Differential Revision: [D46177642](https://our.internmc.facebook.com/intern/diff/D46177642/)

[ghstack-poisoned]

avikchaudhuri committed:
Update on "equality assertions"


Differential Revision: [D46177642](https://our.internmc.facebook.com/intern/diff/D46177642/)

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

avikchaudhuri commented on torch/_export/passes/add_runtime_assertions_for_constraints_pass.py:
```diff
@@ -111,55 +135,81 @@ def placeholder(self, name: str, arg, meta) -> ProxyValue:
+            # If no dynamism is specified, we assume all dimensions are specialized
+            if name not in self.constraints:
+                self._insert_specialized_shapes_assert(arg, all_dims, name, current_inp)
+                return arg
```
umm, need to replace this with `continue`

avikchaudhuri committed:
Update on "equality assertions"


Previously we had runtime asserts for range constraints. This diff adds runtime asserts for equality constraints.

This requires a bit of refactoring that is worth calling out.
1. [Minor] Some of the data structures produced by export and consumed by the runtime assertion pass need to be broadened. This is a WIP. There are some associated code improvements that are included in this diff, but by and large the structures are similar to what exists now. Meanwhile angelayi and I are chatting about how to make it qualitatively better: briefly, we want to index everything by symbols, which are 1-1 with (name, dim) pairs.
2. [Major] The order in which runtime asserts are emitted is changed. Previously we used to do the work in `placeholder`, now this diff adds a hook for "post-processing" after processing of all placeholders is done. This is needed because equality constraints can mention different placeholders. This change also opens the way to optimizing codegen: e.g., each (name, dim) pair should correspond to a single intermediate variable that is reused across runtime asserts. This is future work. 

Differential Revision: [D46177642](https://our.internmc.facebook.com/intern/diff/D46177642/)

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

avikchaudhuri committed:
Update on "equality assertions"


Previously we had runtime asserts for range constraints. This diff adds runtime asserts for equality constraints.

This requires a bit of refactoring that is worth calling out.
1. [Minor] Some of the data structures produced by export and consumed by the runtime assertion pass need to be broadened. This is a WIP. There are some associated code improvements that are included in this diff, but by and large the structures are similar to what exists now. Meanwhile angelayi and I are chatting about how to make it qualitatively better: briefly, we want to index everything by symbols, which are 1-1 with (name, dim) pairs.
2. [Major] The order in which runtime asserts are emitted is changed. Previously we used to do the work in `placeholder`, now this diff adds a hook for "post-processing" after processing of all placeholders is done. This is needed because equality constraints can mention different placeholders. This change also opens the way to optimizing codegen: e.g., each (name, dim) pair should correspond to a single intermediate variable that is reused across runtime asserts. This is future work. 

Differential Revision: [D46177642](https://our.internmc.facebook.com/intern/diff/D46177642/)

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

tugsbayasgalan commented on torch/_export/pass_base.py:
```diff
@@ -294,6 +300,12 @@ def placeholder(self, name: str, arg: Argument, meta: NodeMetadata) -> ProxyValu
         self.tracer.set_metadata(arg_proxy.node, arg)
         return ProxyValue(arg, arg_proxy)
 
+    def postprocess_placeholders(self):
```
Do we expect other passes to use this hook?

avikchaudhuri committed:
Update on "equality assertions"


Previously we had runtime asserts for range constraints. This diff adds runtime asserts for equality constraints.

This requires a bit of refactoring that is worth calling out.
1. [Minor] Some of the data structures produced by export and consumed by the runtime assertion pass need to be broadened. This is a WIP. There are some associated code improvements that are included in this diff, but by and large the structures are similar to what exists now. Meanwhile angelayi and I are chatting about how to make it qualitatively better: briefly, we want to index everything by symbols, which are 1-1 with (name, dim) pairs.
2. [Major] The order in which runtime asserts are emitted is changed. Previously we used to do the work in `placeholder`, now this diff adds a hook for "post-processing" after processing of all placeholders is done. This is needed because equality constraints can mention different placeholders. This change also opens the way to optimizing codegen: e.g., each (name, dim) pair should correspond to a single intermediate variable that is reused across runtime asserts. This is future work. 

Differential Revision: [D46177642](https://our.internmc.facebook.com/intern/diff/D46177642/)

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

avikchaudhuri commented on torch/_export/passes/add_runtime_assertions_for_constraints_pass.py:
```diff
@@ -37,41 +55,47 @@ def __init__(self) -> None:
-                max_vals = [x[1] for x in constraints]
-                min_val = sorted(min_vals, reverse=True)[0]
-                max_val = sorted(max_vals, reverse=False)[0]
+        input_name_to_constraints: Dict[str, List[ConstraintSpec]] = defaultdict(list)
+        for name, dim_constraints in input_name_to_dim_constraints.items():
+            for dim, range_constraints in dim_constraints.ranges.items():
+                if range_constraints:
+                    min_vals, max_vals = zip(*range_constraints)
+                    min_val = max(min_vals)
+                    max_val = min(max_vals)
+                    assert min_val <= max_val
+                    input_name_to_constraints[name].append(
+                        RangeConstraintSpec(dim=dim, min_val=min_val, max_val=max_val)
+                    )
+            for dim, eq_constraints in dim_constraints.equalities.items():
+                for other_name, other_dim in eq_constraints:
+                    input_name_to_constraints[name].append(
+                        EqualityConstraintSpec(dim=dim, other_name=other_name, other_dim=other_dim)
+                    )
```
Yeah, for sure. It should be possible to completely replace the current `input_constraints` and `input_name_to_example_inputs` with the set of data structures we were talking about with symbol / (name, dim) indexing.

I mention `input_name_to_example_inputs` too because the only remaining use of it in the add runtime assertion pass is for specialization constraints.

I'm assuming you will do it, but I could do it here if you want.

avikchaudhuri committed:
Update on "equality assertions"


Previously we had runtime asserts for range constraints. This diff adds runtime asserts for equality constraints.

This requires a bit of refactoring that is worth calling out.
1. [Minor] Some of the data structures produced by export and consumed by the runtime assertion pass need to be broadened. This is a WIP. There are some associated code improvements that are included in this diff, but by and large the structures are similar to what exists now. Meanwhile angelayi and I are chatting about how to make it qualitatively better: briefly, we want to index everything by symbols, which are 1-1 with (name, dim) pairs.
2. [Major] The order in which runtime asserts are emitted is changed. Previously we used to do the work in `placeholder`, now this diff adds a hook for "post-processing" after processing of all placeholders is done. This is needed because equality constraints can mention different placeholders. This change also opens the way to optimizing codegen: e.g., each (name, dim) pair should correspond to a single intermediate variable that is reused across runtime asserts. This is future work. 

Differential Revision: [D46177642](https://our.internmc.facebook.com/intern/diff/D46177642/)

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]
